### ISSUE TITLE
Use @esp.program.modules.base.user_passes_test() to reduce code duplication among program module decorators

### ISSUE THREAD

**Issue by jmoldow** [learning-unlimited/ESP-Website#1322]
`@esp.program.modules.base.user_passes_test()` could possibly be used to define `@meets_deadline`, `@meets_grade`, and all the other program module decorators.

This might require first adding some extra functionality to `@user_passes_test()`. For example, it would need to be able to support manually specifying the error template to use, and be able to support specifying which template variable to use for the error message.

**Comment by imran-techvoyager** [comment]
Hi @jmoldow I’ve been looking into #1322 and wanted to outline a possible refactoring direction.

It seems that several decorators in base.py (e.g., needs_student, needs_teacher, meets_cap, needs_account, meets_grade) share nearly identical scaffolding:

- login check / redirect
- run a predicate
- render a specific error template
- set has_auth_check metadata

However, user_passes_test() currently doesn’t have enough flexibility to fully replace them (e.g., custom templates, extra context, login handling, metadata).

Would you be open to extending user_passes_test() slightly (adding optional parameters for error_template, extra context, and login control), and then incrementally migrating the simplest decorators first?

My thought would be to:
- First enhance user_passes_test in a backwards-compatible way
- Then refactor Tier-1 decorators (needs_student, needs_teacher, meets_cap, needs_account)
- Leave more complex cases like meets_deadline and needs_onsite unchanged for now

@willgearty would love your thoughts as well, especially since this touches Program Modules and auth structure.

If this direction makes sense, I’d be happy to draft a small PR starting with just enhancing user_passes_test().

**Comment by willgearty** [comment]
Yep, that sounds like a reasonable approach to this issue. Thanks @imran-techvoyager 👍
